realm/realm-library/src/objectServer/java/io/realm/internal/network/OkHttpNetworkTransport.java

```diff
@@ -31,47 +34,82 @@ public OkHttpNetworkTransport(@Nullable HttpLogObfuscator httpLogObfuscator) {
         try {
             OkHttpClient client = getClient(timeoutMs);
+
             okhttp3.Response response = null;
             try {
-                Request.Builder builder = new Request.Builder().url(url);
-                switch(method) {
-                    case "get": builder.get(); break;
-                    case "delete": builder.delete(RequestBody.create(JSON, body)); break;
-                    case "patch": builder.patch(RequestBody.create(JSON, body)); break;
-                    case "post": builder.post(RequestBody.create(JSON, body)); break;
-                    case "put": builder.put(RequestBody.create(JSON, body)); break;
-                    default: throw new IllegalArgumentException("Unknown method type: "+ method);
-                }
+                okhttp3.Request request = makeRequest(method, url, headers, body);
 
-                for (Map.Entry<String, String> entry : headers.entrySet()) {
-                    builder.addHeader(entry.getKey(), entry.getValue());
-                }
-                Call call = client.newCall(builder.build());
+                Call call = client.newCall(request);
                 response = call.execute();
                 ResponseBody responseBody = response.body();
                 String result = "";
                 if (responseBody != null) {
                     result = responseBody.string();
                 }
-                return Response.httpResponse(response.code(), parseHeaders(response.headers()), result);
+                return OkHttpResponse.httpResponse(response.code(), parseHeaders(response.headers()), result);
             } catch (IOException ex) {
-                return Response.ioError(ex.toString());
+                return OkHttpResponse.ioError(ex.toString());
             } catch (Exception ex) {
-                return Response.unknownError(ex.toString());
+                return OkHttpResponse.unknownError(ex.toString());
             } finally {
                 if (response != null) {
                     response.close();
                 }
             }
         } catch (Exception e) {
-            return Response.unknownError(e.toString());
+            return OkHttpResponse.unknownError(e.toString());
         }
     }
 
+    @Override
+    public Response sendStreamingRequest(Request request) throws IOException, AppException {
+        OkHttpClient client = getStreamClient();
+
+        okhttp3.Request okRequest = makeRequest(request.getMethod(), request.getUrl(), request.getHeaders(), request.getBody());
+
+        Call call = client.newCall(okRequest);
+        okhttp3.Response response = call.execute();
+
+        if((response.code() >= 300) || ((response.code() < 200) && (response.code() != 0))) {
+            throw new AppException(ErrorCode.fromNativeError(ErrorCode.Type.HTTP, response.code()), "http error code considered fatal");
```

Comment:
Can we forward the underlying http exception message instead? Sometimes the server sends useful information there. "Http error code considered fatal" doesn't really provide anything useful.